### PR TITLE
Replace WeightDecayedAdam with an x10 optimizer for BERT-CoLA

### PR DESCRIPTION
**Pull request by shadaj**
Also includes a bug fix for `TextBatch.paddedAndCollated` to make sure we pad text elements even when there is only one element

### CODE REVIEW

**Comment by texasmichelle** [comment]
Fixes #555

File: Examples/BERT-CoLA/main.swift [hunk 1]
```diff
@@ -86,7 +145,17 @@ for (epoch, epochBatches) in cola.trainingEpochs.prefix(3).enumerated() {
 
         trainingLossSum += loss.scalarized()
         trainingBatchCount += 1
+        gradients.clipByGlobalNorm(clipNorm: 1)
+
+        let step = optimizer.step + 1 // for scheduled rates and bias correction, steps start at 1
+        optimizer.learningRate = scheduledLearningRate(forStep: UInt64(step))
```
Comment (texasmichelle) [inline Examples/BERT-CoLA/main.swift:151]: This is slightly different than the existing implementation, which executes this statement only when `useBiasCorrection` is false, which could explain why the lr decays to zero so quickly. Is your TODO on 113 above in reference to the existing implementation or this implementation with the existing slope of `-5e-7` (rather than `-(2e-5 / (1068 * 3))`)?

File: Examples/BERT-CoLA/main.swift [hunk 2]
```diff
@@ -55,17 +60,65 @@ var cola = try CoLA(
...
-    maxGradientGlobalNorm: 1)
+    TensorVisitorPlan(bertClassifier.differentiableVectorView),
+    defaultOptimizer: makeWeightDecayedAdam(
+      learningRate: 2e-5,
+      beta1: beta1,
+      beta2: beta2
+    )
+)
+
+// TOOD(shadaj): 1068 is a magic number for the number of batches in each epoch, where does this come from?
```
Comment (texasmichelle) [inline Examples/BERT-CoLA/main.swift:112]: I believe this is a function of the number of samples in the training set and the values we're using for `maxSequenceLength` and `batchSize`.

File: Support/Text/TextBatch.swift [hunk 3]
```diff
@@ -71,6 +70,8 @@ extension Collection where Element == TextBatch {
           (before: 0, after: 0),
           (before: 0, after: paddingSize)]))
     }
+
+    if count == 1 { return paddedTexts.first! }
```
Comment (texasmichelle) [inline Support/Text/TextBatch.swift:74]: This is a great catch!

File: Examples/BERT-CoLA/main.swift (hunk 1 above)
Comment (shadaj) [reply Examples/BERT-CoLA/main.swift:151]: Ah didn't notice that, but this opens up a new question. In the original optimizer, the `learningRate` is initialized to 0, so when `useBiasCorrection` is `true` that would result in the `learningRate` always being 0. "Algorithm 1" of the paper seems to indicate that bias correction is just supposed to correct the default `learningRate` in every step instead of accumulating a correction, which seems to match up with this implementation.

File: Examples/BERT-CoLA/main.swift [hunk 4]
```diff
@@ -17,6 +17,10 @@ import Foundation
 import ModelSupport
 import TensorFlow
 import TextModels
+import x10_optimizers_optimizer
+
+// TODO(shadaj): Device.defaultXLA causes CUDA OOM
+let device = Device(kind: .GPU, ordinal: 0, backend: .XLA)
```
[the remark this answers is not in the document]
Comment (shadaj) [reply Examples/BERT-CoLA/main.swift:23]: If I do this, the example crashes on my system since it can't find any device other than the GPU. Ideally, `defaultXLA` would solve this elegantly by picking whichever device is available, but for some reason it causes OOM errors when using the GPU.

**Comment by pschuh** [comment]
This probably only works for x10 because I've merged in the crs into the optimizer. (I think crs might be x10 only). We could make it so that that is optional and doesn't happen when using tfeager.

File: Examples/BERT-CoLA/main.swift [hunk 5]
```diff
@@ -75,8 +128,14 @@ for (epoch, epochBatches) in cola.trainingEpochs.prefix(3).enumerated() {
...
     for batch in epochBatches {
-        let (documents, labels) = (batch.data, Tensor<Float>(batch.label))
-        let (loss, gradients) = valueWithGradient(at: bertClassifier) { model -> Tensor<Float> in
+        let (eagerDocuments, eagerLabels) = (batch.data, Tensor<Float>(batch.label))
+        let documents = TextBatch(
+          tokenIds: Tensor(copying: eagerDocuments.tokenIds, to: device),
+          tokenTypeIds: Tensor(copying: eagerDocuments.tokenTypeIds, to: device),
+          mask: Tensor(copying: eagerDocuments.mask, to: device)
+        )
+        let labels = Tensor(copying: eagerLabels, to: device)
```
[the remark this answers is not in the document]
Comment (shadaj) [reply Examples/BERT-CoLA/main.swift:137]: I don't believe this is possible since on line 131 we convert the original labels (a tensor of `Int`s) to a tensor of `Float`s. As far as I can tell, there's no way to specify the target device of the converted tensor so we have to do a manual move.

[the remark this answers is not in the document]
Comment (texasmichelle) [reply Examples/BERT-CoLA/main.swift:101]: A type conversion should preserve the device. Are you running into errors that indicate otherwise?

File: Examples/BERT-CoLA/main.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (texasmichelle) [reply Examples/BERT-CoLA/main.swift:23]: Something might be wrong with the setup on your system. The easiest way to tell is to try it in Colab, since you can try CPU, GPU, and TPU runtimes (the above works on my system and in colab).

From a [blank notebook](https://colab.research.google.com/notebook#create=true&language=swift), you can add this statement in the first cell to install this branch:
```
%install '.package(url: "https://github.com/shadaj/swift-models", .branch("x10-bert-cola"))' Datasets TextModels ModelSupport
```

File: Examples/BERT-CoLA/main.swift (hunk 1 above)
Comment (texasmichelle) [reply Examples/BERT-CoLA/main.swift:115]: That looks like a bug that was introduced in #544. To inspect the learning rate prior to that, checkout the parent (78dfc05) and look at the value right before [`model.move(along:)`](https://github.com/tensorflow/swift-models/commit/7a401e412561ed2ef8f2efc60d25ae130bd601ca#diff-3f22737c5644cbb02f7e18f73151a659R119) in `WeightDecayedAdam.update()`.

FWIW, we never verified that the original approach was correct, and what you have in this PR looks like an improvement (certainly over the bug!).

File: Examples/BERT-CoLA/main.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (shadaj) [reply Examples/BERT-CoLA/main.swift:101]: If I move to the XLA device earlier, the training takes a significant performance hit. Not sure why though, will work on more performance analysis.

File: Examples/BERT-CoLA/main.swift (hunk 1 above)
Comment (shadaj) [reply Examples/BERT-CoLA/main.swift:115]: Looks like other than the training rate decaying too fast, which was an issue before #544 as well, the new implementation matches the original logic for correction.

File: Examples/BERT-CoLA/main.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (shadaj) [reply Examples/BERT-CoLA/main.swift:100]: Ah, so looks like the source of the issue is that the `CoLA` constructor creates new tensors internally for the labels for each batch, so the device of the constructed labels does not have an effect on the loaded labels for each batch. I think we'll need #583 to fix this.

File: Examples/BERT-CoLA/main.swift [hunk 6]
```diff
@@ -55,17 +60,65 @@ var cola = try CoLA(
```
[the remark this answers is not in the document]
Comment (shadaj) [reply Examples/BERT-CoLA/main.swift:60]: See #583

File: Examples/BERT-CoLA/main.swift [hunk 7]
```diff
@@ -55,17 +60,28 @@ var cola = try CoLA(
...
+      beta2: beta2
+    )
+)
+
+var scheduledLearningRate = LinearlyDecayedParameter(
+  baseParameter: LinearlyWarmedUpParameter(
+      baseParameter: FixedParameter<Float>(2e-5),
+      warmUpStepCount: 10,
+      warmUpOffset: 0),
+  slope: -(2e-5 / (Float(stepsPerBatch) * 3)),  // The LR decays linearly to zero.
```
Comment (texasmichelle) [inline Examples/BERT-CoLA/main.swift:82]: Good call creating a variable, although we could probably calculate this somehow instead. Optional: move 3 into `epochCount` (also L87) and 2e-5 into `learningRate` (L71, L79).

**Comment by 8bitmp3** [comment]
> This is an incredible contribution for your SECOND DAY on the team 🎉 Amazing!!

I missed the news @texasmichelle . Congrats on joining the team @shadaj ! This PR is awesome 👍 Let me know if you have time to answer a few questions. If you have time can you go over if: 
1) it follows closely the TF implementation of "Adam4BERT" ;
2) there is no bias compensation similar to `BertAdam` by PyTorch, which is based on TensorFlow's Adam optimizer (I read about it on HuggingFace's repo) (assuming the weight decay thing is fixed, just like in `BertAdam`); and 
3) the X10 optimizer version considers the [memory](https://github.com/google-research/bert) efficiency vs performance/results trade-off that the original Google work mentioned on GitHub? 

Also, is using the name `x10_optimizers_optimizer`—especially in the `import LibraryName` context—permanent and are you open to naming suggestions? Naming things is hard 🙂 (cc @saeta).

Sorry if this has been covered before and I missed it. Cheers!
